# Importing Libraries

In [1]:
# Importing libraries (pandas, numpy, os)
import pandas as pd
import numpy as np
import os

In [2]:
#Importing data set products.csv
df_prods = pd.read_csv(r'/Users/shailioza/Documents/CareerFoundry Projects/Instacart Basket Analysis/02 Data/Original Data/products.csv', index_col = False)

In [3]:
#Importing data set orders_wrangled.csv
df_ords = pd.read_csv(r'/Users/shailioza/Documents/CareerFoundry Projects/Instacart Basket Analysis/02 Data/Prepared Data/orders_wrangled.csv', index_col = False)

In [4]:
# 2.Run the df.describe() function on your df_ords dataframe. 
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [5]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Min value 1.000000, Max value 99999.000000, and Mean 9.994, which is very close to the minimum value.

In [6]:
# 3.Check for mixed-type data in your df_ords dataframe:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
        print (col, ' mixed')
  else: print(col, ' consistent')

Unnamed: 0  consistent
order_id  consistent
user_id  consistent
order_number  consistent
orders_day_of_week  consistent
order_hour_of_day  consistent
days_since_prior_order  consistent


In [7]:
# Check for mixed-type columns in df_prods
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col, ' mixed')
  else: print(col, ' consistent')

product_id  consistent
product_name  mixed
aisle_id  consistent
department_id  consistent
prices  consistent


4. No mixed-type data found

In [8]:
# 5.Run a check for missing values in your df_ords dataframe:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [9]:
# 5.Run a check for missing values in your df_prods dataframe:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

There are 206209 missing values in the column, 'days_since_prior_order'

In [10]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [11]:
df_nan1 = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
df_nan1

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
df_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [14]:
df_ords['user_id'] = df_ords['user_id'].astype('int64')
df_ords.user_id.max()

206209

The number of missing values is the same as the number of user ids in the data frame. This could mean that these are the ids of new customers who obviously do not have a previous order, which is why those variables are empty, or missing. 

6.Address the missing values using an appropriate method.

I have chosen to leave the missing values as it is, because they signify the number of new customers, not missing data. This information could also be useful in analysis in some way in the future. 

In [15]:
df_prods['product_id'] = df_prods['product_id'].astype('int64')
df_prods.product_id.max()

49688

In [16]:
# Create df without nan values:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
# 7.Run a check for duplicate values in your df_ords data
df_ords_dups = df_ords[df_ords.duplicated()]

In [19]:
# 7.Run a check for duplicate values in your df_prods data
df_prods_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates found

In [21]:
df_prods_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
# Creat dropped dups list 
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [23]:
df_prods_clean_no_dups.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

In [25]:
# 9.Export changes
df_ords.to_csv(r'/Users/shailioza/Documents/CareerFoundry Projects/Instacart Basket Analysis/02 Data/Prepared Data/orders_checked.csv')

In [26]:
orders_checked

NameError: name 'orders_checked' is not defined

In [59]:
# Export changes df:
df_prods_clean_no_dups.to_csv(r'/Users/shailioza/Documents/CareerFoundry Projects/Instacart Basket Analysis/02 Data/Prepared Data/products_checked.csv')